### ISSUE TITLE
PerlinNoise() requires square matrices?

### ISSUE THREAD

**Issue by tpoisot** [EcoJulia/NeutralLandscapes.jl#33]
~~~
julia> rand(PerlinNoise(), 10, 10);

julia> rand(PerlinNoise(), 10, 12);
ERROR: DimensionMismatch("array could not be broadcast to match destination")
~~~

**Comment by rafaqz** [comment]
Oops

**Comment by rafaqz** [comment]
Also apologies I'm busy and on the road at the moment, so I'm missing all the fun. But should have some time for this in a few days.

**Comment by mkborregaard** [comment]
@gottacatchenall used a solution where he allocated a temporary square array big enough to hold the non-square one - might make sense to choose the same approach in the two places?

**Comment by tpoisot** [comment]
@mkborregaard I agree - @gottacatchenall is this already a function that @rafaqz can re-use?

**Comment by gottacatchenall** [comment]
not quite, i check the size via 
```
function _landscape!(mat, alg::Union{DiamondSquare, MidpointDisplacement}; kw...) where {IT <: Integer}
    rightSize::Bool = _isPowerOfTwo(size(mat)[1]-1) && _isPowerOfTwo(size(mat)[2]-1)

    dsMat = mat
    if !rightSize
        dim1, dim2 = size(mat)
        smallestContainingLattice::Int = 2^ceil(log2(max(dim1, dim2))) + 1
        @warn "$alg cannot be run on the input dimensions ($dim1 x $dim2),
                and will instead run on the next smallest valid size ($smallestContainingLattice x $smallestContainingLattice).
                This can slow performance as it involves additional memory allocation."
        dsMat = zeros(smallestContainingLattice, smallestContainingLattice)
    end
    _diamondsquare!(dsMat, alg)

    mat .= dsMat[1:size(mat)[1], 1:size(mat)[2]]
end
```
which enforeces it must be square but has the additional condition that the size must be `(2^n)+1`, although this could be implemented in to something more general like

```
function _landscape!(mat, alg)
  matPointer = mat
  if !rightSize(mat, alg) 
     # allocation correct size matrix to matPointer
  end 
  mat[:] .= matPointer[1:size(mat)[1], 1:size(mat)[2]] 
end
```

**Comment by rafaqz** [comment]
Thats what happens already, I just messed it up somehow

**Comment by tpoisot** [comment]
This was closed via faaeed5120d1c00
